superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+that we can assume for some level of cross-L2 guarantee, but are not otherwise safe to use yet.
+
+### Messaging Shreds
+
+The idea of streaming partial blocks can be implemented as extension of the above described append-log model;
+the tip of the append-log is maintained by the shreds, and does not have to be fetched anymore.
+This can happen in-memory, with periodic flushes to disk.
+
+### Messaging "Horizon"
+
+When prefetching data, any individual L2 chain must not go too far ahead before other L2s catch up,
+to keep work balanced and verifiable. The "horizon" is the furthest (into future) timestamp that can be seen,
+and can be entirely verified with data up to that point, thanks to the timestamp invariant.
+
+Using this "horizon" property, nodes can implement a form of back-pressure,
+where it is known exactly whether or not all data has been retrieved to verify remaining open cross-L2 dependencies.
+
+### Message safety is not different than block safety
+
+Consider the following block safety levels, and their messaging equivalents:
+
+```
+enum SafetyLevel {
+    INVALID, # conflicting/invalidated data, to be removed from the canonical view
+    UNSAFE,  # not cross-verified anything, yet.
+    CROSS_UNSAFE,  # cross-verified L2
+    SAFE,  # verified L1 also
+    FINAL,  # irreversible
+}
+```
+
+When considering the global log of interop hashes per L2 chain,
+the safety level label just moves along that log, like it does on the chain of blocks.
+
+### Simple interface:
+
+The verifier side is really all just one interface:
+
+```
+type MessageVerifier interface {
+    Check(Identifier, PayloadHash) SafetyLevel
+}
+```
+
+This can be backed by:
+
+- Wrapper around an ethereum RPC:
+    - using `eth_getLogs`, fetch log with 1-block 1-contract narrow scope, check if it matches,
+      return safety-level of block it was included in.
+    - using `eth_getTransactionReceipt` (if tx hash is known), fetch logs of specific tx, rely on metadata for event-index,
+      return safety-level of block it was included in.
+    - using `eth_getBlockReceipts`, fetch all logs of block, check the specific log,
+      return safety-level of block it was included in.
+- Wrapper around a append-log "database", as described in the ideas above:
+  search the database for the data, check against computed message-hash, and then return safety level.
+- Wrapper around a shreds reader: if we maintain the tip of the chain in memory based on streamed data,
+  we can serve queries for that data, and otherwise fallback to another source.
+
+In a fault-proof, a trustless version of the `eth_getBlockReceipts` can support the query,
+as that verifies the log-index correctness relative to the start of the block, and inclusion-proof to the block etc.
+
+### Batching
+
+We may want to consider batch-queries as part of the interface;
+query `[]BatchElem` (`BatchElem = (Identifier, PayloadHash)`), return `[]SafetyLevel`.
```

Comment:
You are right that its an optimization to support this directly in the interface rather than do a batch RPC call